politeiawww/wsdcrdata.go

```diff
@@ -26,13 +30,46 @@ var (
 	// errSubNotFound is emitted when attempting to unsubscribe to an
 	// event that has not yet been subscribed to.
 	errSubNotFound = errors.New("subscription not found")
+
+	errShutdown = errors.New("ws connection is shut down")
 )
 
 // wsDcrdata is the context used for managing a dcrdata websocket connection.
 type wsDcrdata struct {
 	sync.RWMutex
+	shutdown      bool
 	client        *client.Client      // dcrdata websocket client
 	subscriptions map[string]struct{} // Client subscriptions
+	recvMsgChan   chan *client.ClientMessage
```

Comment:
It seems like you made this unnecessarily complicated.  Why do you need this channel?

Why not just check isShutdown() at the beginning of any method that makes an external call (subToAddr, reconnect, etc) and return errShutdown if true.

Reply:
I added this to let listeners know that wsDcrdata is being shutdown so they can gracefully stop listening. Otherwise, the channel will output a null message and the listener will attempt to reconnect (and fail since isShutdown() is checked in the beginning of the reconnect method). Nothing bad will happen without this, other than a spew of error messages when there is actually no error.

Reply:
Ok, I see what you're saying.  The correct way to do this would be to get rid of the channel and have a wsDcrdata method for all of the external calls that are being made (ping, subscribe).  Those methods will check isShutdown before making the call.  

You'd need to replace all of the direct calls with the wsDcrdata methods.  Ex: replace all the `w.client.Subscribe(event)` with `w.Subscribe(event)`.

Make sense?

Reply:
Yeah makes sense.